File: dataStructures/Heap/binaryHeap.js
class BinaryHeap {
    constructor (data= [], comparer = (a, b) => +a < +b) {
        this._data = data.slice();
        this._comparer = comparer;

        this._siftDown(0);
    }

    _swap(i, j) {
        [this._data[i], this._data[j]] = [this._data[j], this._data[i]];
    }

    _siftUp(i) {
        while (this._comparer(this._data[i], this._data[((i - 1) >> 1)])) {
            this._swap(i, ((i - 1) >> 1));
            i = ((i - 1) >> 1);
        }
    }

    _siftDown(i) {
        while (this.getSize() > (i << 1) + 1) {
            const left = (i << 1) + 1;
            const right = (i << 1) + 2;
            let j = left;

            if (this.getSize() > right && this._comparer(this._data[right], this._data[left])) {
                j = right;
            }

            if (this._data[i] === this._data[j] || this._comparer(this._data[i], this._data[j])) {
                break;
            }

            this._swap(i, j);
            i = j;
        }
    }

    getSize() {
        return this._data.length;
    }

    insert(a) {
        this._data.push(a);
        this._siftUp(this.getSize() - 1);
    }

    extractRoot() {
        const root = this._data.shift();

        if (this.getSize()) {
            this._data.unshift(this._data.pop());
            this._siftDown(0);
        }

        return root;
    }
}

module.exports = BinaryHeap;

File: dataStructures/Heap/binaryHeap.cs
private class MinComparer : IComparer<int>
{
    public int Compare(int x, int y) => x.CompareTo(y);
}

private class MaxComparer : IComparer<int>
{
    public int Compare(int x, int y) => y.CompareTo(x);
}

private class Heap<T>
{
    private readonly IList<T> _data;
    private readonly IComparer<T> _comparer;

    public int Count => _data.Count;
    public T Top => _data[0];

    public Heap(IList<T> inputs, IComparer<T> comparer = null)
    {
        _comparer = comparer ?? Comparer<T>.Default;
        _data = inputs;
        for (int i = Count / 2; i >= 0; i--)
        {
            SiftDown(i);
        }
    }

    public Heap(IEnumerable<T> inputs, IComparer<T> comparer = null) : this(inputs.ToList(), comparer)
    {
    }

    private void Swap(int i, int j)
    {
        var tmp = _data[i];
        _data[i] = _data[j];
        _data[j] = tmp;
    }

    private void SiftDown(int i)
    {
        while (2 * i + 1 < _data.Count)
        {
            int left = 2 * i + 1;
            int right = 2 * i + 2;
            int j = left;

            if (right < _data.Count && _comparer.Compare(_data[right], _data[left]) < 0)
            {
                j = right;
            }

            if (_comparer.Compare(_data[i], _data[j]) <= 0)
            {
                break;
            }

            Swap(i, j);
            i = j;
        }
    }

    private void SiftUp(int i)
    {
        while (_comparer.Compare(_data[i], _data[(i - 1) / 2]) < 0)
        {
            Swap(i, (i - 1) / 2);
            i = (i - 1) / 2;
        }
    }

    public T ExtractTop()
    {
        T top = Top;
        _data[0] = _data.Last();
        _data.RemoveAt(Count - 1);
        SiftDown(0);
        return top;
    }

    public void Add(T value)
    {
        _data.Add(value);
        SiftUp(Count - 1);
    }
}
